File: backend/src/main/java/apirest/workzen/service/FichajeService.java
package apirest.workzen.service;

import java.util.List;

import org.springframework.security.core.Authentication;

import apirest.workzen.modelo.entities.Fichaje;

/**
 * Interfaz de servicio para la gestion de fichajes.
 */
public interface FichajeService extends ICrudGenerico<Fichaje, Integer> {

    /**
     * Obtiene el fichaje activo del empleado autenticado.
     *
     * @param auth autenticacion del usuario
     * @return el fichaje activo o null si no existe
     */
    Fichaje obtenerFichajeActivo(Authentication auth);

    /**
     * Inicia un nuevo fichaje para el empleado autenticado.
     *
     * @param auth autenticacion del usuario
     * @return el fichaje creado, o null si ya existe uno activo o hay error
     */
    Fichaje iniciarFichaje(Authentication auth);

    /**
     * Crea una pausa en el fichaje indicado.
     *
     * @param auth autenticacion del usuario
     * @param fichajeId id del fichaje
     * @return el fichaje actualizado, o null si hay error
     */
    Fichaje pausarFichaje(Authentication auth, int fichajeId);

    /**
     * Reanuda el fichaje cerrando la pausa activa.
     *
     * @param auth autenticacion del usuario
     * @param fichajeId id del fichaje
     * @return el fichaje actualizado, o null si hay error
     */
    Fichaje reanudarFichaje(Authentication auth, int fichajeId);

    /**
     * Finaliza el fichaje estableciendo fecha de salida y estado finalizado.
     *
     * @param auth autenticacion del usuario
     * @param fichajeId id del fichaje
     * @return el fichaje finalizado, o null si hay error
     */
    Fichaje finalizarFichaje(Authentication auth, int fichajeId);

    /**
     * Obtiene los fichajes de un empleado filtrados por mes y año.
     *
     * @param empleadoId id del empleado
     * @param mes        mes (1-12)
     * @param anio       año
     * @return lista de fichajes del mes
     */
    List<Fichaje> obtenerPorEmpleadoYMes(int empleadoId, int mes, int anio);

    /**
     * Obtiene todos los fichajes filtrados por mes y año (admin).
     *
     * @param mes  mes (1-12)
     * @param anio año
     * @return lista de todos los fichajes del mes
     */
    List<Fichaje> obtenerTodosPorMes(int mes, int anio);
}

File: frontend/src/app/services/fichaje-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

/** Pausa dentro de un fichaje */
export interface Pausa {
  id: number;
  horaInicio: string;
  horaFin: string | null;
}

/** Fichaje del empleado con sus pausas */
export interface Fichaje {
  id: number;
  fechaEntrada: string;
  fechaSalida: string | null;
  estado: string;
  pausas: Pausa[];
}

@Injectable({
  providedIn: 'root',
})
export class FichajeService {
  private apiUrl = `${environment.apiUrl}/fichajes`;

  constructor(private http: HttpClient) {}

  /** Obtener el fichaje activo del empleado (o 204 si no hay) */
  getFichajeActivo(): Observable<Fichaje> {
    return this.http.get<Fichaje>(`${this.apiUrl}/activo`);
  }

  /** Iniciar un nuevo fichaje */
  iniciarFichaje(): Observable<Fichaje> {
    return this.http.post<Fichaje>(`${this.apiUrl}/iniciar`, {});
  }

  /** Pausar el fichaje activo */
  pausarFichaje(id: number): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}/pausar`, {});
  }

  /** Reanudar el fichaje (finalizar pausa activa) */
  reanudarFichaje(id: number): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}/reanudar`, {});
  }

  /** Finalizar el fichaje */
  finalizarFichaje(id: number): Observable<Fichaje> {
    return this.http.put<Fichaje>(`${this.apiUrl}/${id}/finalizar`, {});
  }

  /** Obtener mis fichajes por mes y anio */
  getMisFichajes(mes: number, anio: number): Observable<Fichaje[]> {
    return this.http.get<Fichaje[]>(`${this.apiUrl}/mis-fichajes?mes=${mes}&anio=${anio}`);
  }

  /** Obtener todos los fichajes por mes y anio (admin) */
  getTodosFichajes(mes: number, anio: number): Observable<FichajeAdmin[]> {
    return this.http.get<FichajeAdmin[]>(`${this.apiUrl}/todos?mes=${mes}&anio=${anio}`);
  }
}

/** Fichaje con datos del empleado para vista admin */
export interface FichajeAdmin extends Fichaje {
  empleadoId: number;
  empleadoNombre: string;
}
